File: src/realign.py
import pysam
import argparse
import sys
import re
import parasail
from intervaltree import Interval, IntervalTree
from collections import defaultdict

scoring_matrix = parasail.matrix_create("ACGT", 5, -1)


def get_id(header, name):
    i = 0
    while i < len(header['SQ']):
        if header['SQ'][i]['SN'] == name:
            return i
        else:
            i += 1
    return 0


def get_hap_set(chrom, start, window, sample_to_fastq, sample_to_bam, ref_fa, insert_list, ref_window):
    haplotype_sequences = defaultdict(list)
    reads_per_sample = defaultdict(list)
    read_sequences = defaultdict(str)
    best_haplotype_per_read= defaultdict(list)
    insert_reads = {}
    for sample in sample_to_bam:
        in_bam = pysam.AlignmentFile(sample_to_bam[sample])
        sam_reader = in_bam.fetch(region=chrom+":"+str(start)+"-"+str(start+window))
        fq = pysam.FastaFile(filename = sample_to_fastq[sample])
        for record in sam_reader:
            # Grab read sequences from fastq
            read_sequence = fq.fetch(record.query_name)
            read_sequences[record.query_name+"\t"+sample] = read_sequence
    # Evaluate the base haplotype 
    base_ref_seq = ref_fa.fetch(region=chrom+":"+str(start-ref_window)+"-"+str(start+ref_window))
    haplotype_sequences["base"] = [base_ref_seq, 0, ""]
    for insert in insert_list:
        # Compute where the insert sequence should be spliced in
        avg_insert_pos = int((int(insert.data[0][1]) + int(insert.data[0][2]))/2)
        insert_pos = ((avg_insert_pos - start) + ref_window)
        haplotype_sequences[insert.data[0][3]] =  [base_ref_seq[:insert_pos]+insert.data[0][7]+base_ref_seq[insert_pos:], insert_pos - ref_window, insert.data[0][7]]
        insert_reads[insert.data[0][3]] = 1
    # Once we have haplotype set, align each read to each haplotype and mark wich one was best scoring
    haplotype_profiles = {}
    for name in haplotype_sequences:
        haplotype_profiles[name] = parasail.profile_create_16(haplotype_sequences[name][0], scoring_matrix)
    for item in read_sequences:
        read = item.split('\t')[0]
        if read not in insert_reads:
            continue
        sample = item.split('\t')[1]
        #print(read)
        read_sequence = read_sequences[read+"\t"+sample]
        for name in haplotype_sequences:
            result = parasail.sw_striped_profile_16(haplotype_profiles[name], read_sequence, 5, 4)
            best_haplotype_per_read[read].append([result.score, name, read, sample])
            #print([result.score, name, read, sample])
    return (best_haplotype_per_read, haplotype_sequences, haplotype_profiles, read_sequences)

def get_realign_window(haplotype_profiles, read_sequences):
    # Get all reads in region per bam
    best_haplotype_per_read = defaultdict(list)
    for item in read_sequences:
        read = item.split('\t')[0]
        sample = item.split('\t')[1]
        # Grab read sequences from fastq
        #print("realign "+item)
        read_sequence = read_sequences[item]
        # Align each read to best haplotypes
        best_score = 0
        best_hap = ""
        for hap in haplotype_profiles:
            result = parasail.sw_striped_profile_16(haplotype_profiles[hap], read_sequence, 5, 4)
            if result.score > best_score:
                best_score = result.score
                best_hap = hap
        best_haplotype_per_read[item] = [best_hap, result]
    return best_haplotype_per_read


def get_insert_pos(result, insert_ref_start_pos, insert_length):
    insert_ref_end_pos = insert_ref_start_pos + insert_length
    cigar = result.cigar.decode.decode('utf-8')
    # Ref seq is the haplotype, query is the read
    cgs = re.findall('[0-9]*[A-Z=]', cigar)
    ref_hap_count = 0
    read_count = 0
    start_pos = 0
    end_pos = 0
    for cg in cgs:
        if cg.endswith("M"):
            ref_hap_count += int(cg[:cg.find("M")])
            read_count += int(cg[:cg.find("M")])
        elif cg.endswith("I"):
            read_count += int(cg[:cg.find("I")])
        elif cg.endswith("D"):
            ref_hap_count += int(cg[:cg.find("D")])
        elif cg.endswith("N"):
            ref_hap_count += int(cg[:cg.find("N")])
        elif cg.endswith("S"):
            read_count += int(cg[:cg.find("S")])
        elif cg.endswith("="):
            ref_hap_count += int(cg[:cg.find("=")])
            read_count += int(cg[:cg.find("=")])
        elif cg.endswith("X"):
            ref_hap_count += int(cg[:cg.find("X")])
            read_count += int(cg[:cg.find("X")])
        if ref_hap_count <= insert_ref_start_pos:
            start_pos = read_count
        if ref_hap_count >= insert_ref_end_pos and end_pos == 0:
            end_pos = read_count
    return(start_pos, end_pos)

def get_start_end_pos(result):
    cigar = result.cigar.decode.decode('utf-8')
    # Ref seq is the haplotype, query is the read
    cgs = re.findall('[0-9]*[A-Z=]', cigar)
    ref_hap_count = 0
    read_count = 0
    start_pos = 0
    end_pos = 0
    for cg in cgs:
        if cg.endswith("M"):
            ref_hap_count += int(cg[:cg.find("M")])
            read_count += int(cg[:cg.find("M")])
        elif cg.endswith("I"):
            read_count += int(cg[:cg.find("I")])
        elif cg.endswith("D"):
            ref_hap_count += int(cg[:cg.find("D")])
        elif cg.endswith("N"):
            ref_hap_count += int(cg[:cg.find("N")])
        elif cg.endswith("S"):
            read_count += int(cg[:cg.find("S")])
        elif cg.endswith("H"):
            read_count += int(cg[:cg.find("H")])
        elif cg.endswith("="):
            ref_hap_count += int(cg[:cg.find("=")])
            read_count += int(cg[:cg.find("=")])
        elif cg.endswith("X"):
            ref_hap_count += int(cg[:cg.find("X")])
            read_count += int(cg[:cg.find("X")])
        if read_count == 0:
            start_pos = ref_hap_count
        end_pos = ref_hap_count
    return(start_pos, end_pos)

def get_start_pos(result):
    cigar = result.cigar.decode.decode('utf-8')
    # Ref seq is the haplotype, query is the read
    cgs = re.findall('[0-9]*[A-Z=]', cigar)
    ref_hap_count = 0
    read_count = 0
    start_pos = 0
    count = 0
    for cg in cgs:
        if cg.endswith("M"):
            ref_hap_count += int(cg[:cg.find("M")])
            read_count += int(cg[:cg.find("M")])
            count += 1
        elif cg.endswith("I"):
            read_count += int(cg[:cg.find("I")])
            count += 1
        elif cg.endswith("D"):
            ref_hap_count += int(cg[:cg.find("D")])
            if count == 0:
                # Start on a deletion
                start_pos = ref_hap_count
            count += 1
        elif cg.endswith("N"):
            ref_hap_count += int(cg[:cg.find("N")])
            count += 1
        elif cg.endswith("S"):
            read_count += int(cg[:cg.find("S")])
            count += 1
        elif cg.endswith("H"):
            read_count += int(cg[:cg.find("H")])
            count += 1
        elif cg.endswith("="):
            ref_hap_count += int(cg[:cg.find("=")])
            read_count += int(cg[:cg.find("=")])
            count += 1
        elif cg.endswith("X"):
            ref_hap_count += int(cg[:cg.find("X")])
            read_count += int(cg[:cg.find("X")])
            count += 1
    cleaned_cigar = "".join(cgs[1:len(cgs) - 1])
    return(start_pos, cleaned_cigar)

def get_best_haplotypes(best_haplotype_per_read, haplotype_profiles, haplotype_sequences):
    best_haplotype_profiles = {}
    best_haplotype_profiles["base"] = haplotype_profiles["base"]
    # For each read get set of haps that are within 5% of the best to account for small scoring variations
    top_set = defaultdict(list)
    for read in best_haplotype_per_read:
        best_haplotype_per_read[read] = sorted(best_haplotype_per_read[read], key=lambda x: x[0], reverse=True)
        max_score = -1
        for item in best_haplotype_per_read[read]:
            if item[1] == "base":
                continue
            if max_score == -1:
                max_score = item[0]
                continue
            if float(item[0])/max_score > 0.95:
                top_set[read].append(item[1])
    # Top set should have all the best haplotypes for each read
    reads = list(top_set.keys())
    seen = {}
    groups = defaultdict(list)
    count = 0
    for i in range(len(reads)):
        if i not in seen:
            seen[i] = 1
            groups[i].append(reads[i])
        for j in range(len(reads)):
            if i >= j:
                continue
            if j in seen:
                continue
            if top_set[reads[i]] == top_set[reads[j]]:
                seen[j] = 1
                groups[i].append(reads[j])
    # Have the groups 
    for i in groups:
        best_haps = top_set[groups[i][0]]
        if len(best_haps) == 0:
            # The base was the best hap only, Its already been added to the set
            pass
        else:
            # Select the longest hap from the best haps list
            max_len = 0
            max_hap = ""
            for hap in best_haps:
                if len(haplotype_sequences[hap]) > max_len:
                    max_len = len(haplotype_sequences[hap])
                    max_hap = hap
            best_haplotype_profiles[max_hap] = haplotype_profiles[max_hap]
    return best_haplotype_profiles

# Assumes that the bam_list and tsv_list line up, will throw an error if they don't
def realign_candidate_insertions(bam_list, tsv_list, fastq_list, output_tsv, output_bam, window, ref, ref_window):
    # Parse bam and tsv lists. Check if they are same size
    bams = bam_list.split(',')
    tsvs = tsv_list.split(',')
    fastqs = fastq_list.split(',')
    assert len(bams) == len(tsvs), "The length of the tsv list and bam list should match"
    assert len(bams) == len(fastqs), "The length of the fastq list and bam list should match"
    sample_to_fastq = defaultdict(str)
    sample_to_tsv = defaultdict(str)
    sample_to_bam = defaultdict(str)
    for i in range(len(bams)):
        path = bams[i].split('/')
        sample = path[len(path)-1].split(".bam")[0]
        sample_to_tsv[sample] = tsvs[i]
        sample_to_bam[sample] = bams[i]
        sample_to_fastq[sample] = fastqs[i]
    header = pysam.AlignmentFile(bams[0]).header
    # Read in insertions. Only consider passing inserts
    inserts_per_chrom = defaultdict(IntervalTree)
    min_pos_per_chrom = defaultdict(int)
    max_pos_per_chrom = defaultdict(int)
    for sample in sample_to_tsv:
        tsv = sample_to_tsv[sample]
        with open(tsv, 'r') as in_tsv:
            count = 0
            for line in in_tsv:
                if count == 0:
                    count = 1
                    continue
                row = line.split('\t')
                if row[8] != "PASS":
                    continue
                chrom = row[0]
                inserts_per_chrom[chrom][int(row[1]):int(row[2])] = (row, sample)
                if min_pos_per_chrom[chrom] == 0:
                    min_pos_per_chrom[chrom] = int(row[1])
                if int(row[1]) < min_pos_per_chrom[chrom]:
                    min_pos_per_chrom[chrom] = int(row[1])
                if int(row[1]) > max_pos_per_chrom[chrom]:
                    max_pos_per_chrom[chrom] = int(row[1])
    # Re-align each insertion on its own first. If the insert does not map end to end then it may be off
    # We will align these insertions to other passinng inserts that do align end to end in a later step
    with open(output_tsv, 'w') as out_tsv:
        out_tsv.write("Chromosome\tStart\tEnd\tAltHaplotypeRead\tInsertSequence\tSupportingReads\n")
        with pysam.AlignmentFile(output_bam, "wb", header=header) as outf:
            with pysam.FastaFile(filename=ref) as ref_fa:
                # Go over each window starting at min pos
                for chrom in min_pos_per_chrom:
                    chrom_seq = ref_fa.fetch(chrom)
                    for i in range(min_pos_per_chrom[chrom], max_pos_per_chrom[chrom]+window, window):
                        insert_list = inserts_per_chrom[chrom][i:i+window]
                        insert_reads = {}
                        if len(insert_list) == 0:
                            #print("No inserts at "+ str(i))
                            continue
                        print(i, file=sys.stderr)
                        for item in insert_list:
                            insert_reads[item.data[0][3]+"\t"+item.data[1]] = 1
                        # Generate set of alternative haplotypes by splicing in insert sequence
                        best_haplotype_per_read, haplotype_sequences, haplotype_profiles, read_sequences = get_hap_set(chrom, i, window, sample_to_fastq, sample_to_bam, ref_fa, insert_list, ref_window)
                        # Compute best mapping haplotype set
                        best_haplotype_profiles = get_best_haplotypes(best_haplotype_per_read, haplotype_profiles, haplotype_sequences)
                        # Map all reads to the set of best haplotype sequences
                        reads_in_window_to_haplotypes = get_realign_window(best_haplotype_profiles, read_sequences)
                        reads_per_haplotype = defaultdict(list)
                        # Get list of reads that support each haplotype in reads_in_window_to_halpotypes
                        for item in reads_in_window_to_haplotypes:
                            read = item.split('\t')[0]
                            sample = item.split('\t')[1]
                            hap = reads_in_window_to_haplotypes[item][0]
                            reads_per_haplotype[hap].append([read,sample])
                        # For each hap compute the TSV record. Include insertion sequence and list of supporting reads
                        for hap in reads_per_haplotype:
                            if hap == "base":
                                continue
                            # First get the positions of the main entry 
                            tsv_row = [chrom, str(i+haplotype_sequences[hap][1]), str(i+haplotype_sequences[hap][1]+1), hap+'_insert_hap', haplotype_sequences[hap][2]]
                            # Then get the positions for each supporting read baed on the re-alignment for that read
                            reads = []
                            for item in reads_per_haplotype[hap]:
                                # Get the re-alignmnet record and compute the position of the insertion based on parsing the cigar string
                                #result = reads_in_window_to_haplotypes[item[0]+'\t'+item[1]][1]
                                #read_sequence = read_sequences[item[0]+'\t'+item[1]]
                                #upstream_result = parasail.sw_trace_striped_profile_16(upstream_profile, read_sequence, 5, 4)
                                #downstream_result = parasail.sw_trace_striped_profile_16(downstream_profile, read_sequence, 5, 4)
                                #start_pos, end_pos = get_start_end_pos(result, haplotype_sequences[hap][1] + ref_window, len(haplotype_sequences[hap][2]))
                                reads.append(item[0]+'_'+item[1])
                            tsv_row.append(','.join(reads))
                            out_tsv.write('\t'.join(tsv_row)+'\n')
                            # Can then generate a BAM record for each haplotype sequence mapped to the reference
                            # Need to Re-Map haplotype sequence to the reference to get cigar string
                            result = parasail.sg_trace_striped_16(haplotype_sequences[hap][0], chrom_seq[i-len(haplotype_sequences[hap][0]):i+len(haplotype_sequences[hap][0])], 5, 4, scoring_matrix)
                            start_pos, cleaned_cigar = get_start_pos(result)
                            ###################################################################################
                            # Use Different CIGAR string as parasial may not align insert as single insertion #
                            ###################################################################################
                            cleaned_cigar = str(ref_window)+"M"+str(len(haplotype_sequences[hap][2]))+"I"+str(ref_window)+"M"
                            # Create alignment record for the haplotype based on CIGAR string
                            start_pos = i-len(haplotype_sequences[hap][0]) + start_pos
                            a = pysam.AlignedSegment(header=header)
                            a.query_name = hap+'_insert_hap'
                            a.flag = 0
                            a.reference_id = get_id(header, chrom)
                            a.reference_start = start_pos
                            a.mapping_quality = 60
                            a.cigarstring = cleaned_cigar
                            a.query_sequence = haplotype_sequences[hap][0]
                            outf.write(a)

File: src/ReAlign.cpp
#include <iostream>
#include <unistd.h>
#include <sstream>
#include <fstream>
#include <algorithm>
#include <limits>
#include <cmath> 
#include "ReAlign.hpp"
#include "TsvRecord.hpp"
#include <stdexcept>
#include "htslib/vcf.h"
#include "htslib/sam.h"
#include "htslib/faidx.h"
#include "parasail.h"

struct AlignmentResult {
    int score;
    std::string probe_name;
    std::string query;
    std::string ref;
    std::string comp;
    std::string oligo;
    char orientation;
    char* cigar;
    unsigned char* alignment;
    int* endLocation;
    int alignmentLength;
};

std::string dna_reverse_complement(std::string seq) {
    reverse(seq.begin(),seq.end());
    for (std::size_t i = 0; i < seq.length(); ++i){
        switch (seq[i]){
        case 'A':
            seq[i] = 'T';
            break;
        case 'C':
            seq[i] = 'G';
            break;
        case 'G':
            seq[i] = 'C';
            break;
        case 'T':
            seq[i] = 'A';
            break;
        }
    }
    return seq;
}

parasail_matrix_t *matrix = parasail_matrix_create("ACGT", 5, -1);

inline AlignmentResult align_parasail(std::string reads, std::string control, parasail_matrix_t *matrix) {
    AlignmentResult complete_result;
    std::string reverse = dna_reverse_complement(reads);
    
    parasail_result_t* result = parasail_sg_trace_scan_16(reads.c_str(),reads.length(),control.c_str(),control.length(),11,1,matrix);
    parasail_traceback_t* traceback = parasail_result_get_traceback(result,reads.c_str(), reads.length(), control.c_str(), control.length(),matrix,'|','*','*');
    parasail_result_t* result_reverse = parasail_sg_trace_scan_16(reverse.c_str(),reverse.length(), control.c_str(), control.length(),11,1,matrix);
    parasail_traceback_t* traceback_reverse = parasail_result_get_traceback(result_reverse,reverse.c_str(), reverse.length(), control.c_str(), control.length(),matrix,'|','*','*');
    parasail_cigar_t* cigar = result->score > result_reverse->score ? parasail_result_get_cigar(result, reads.c_str(), reads.length(), control.c_str(), control.length(), matrix) : parasail_result_get_cigar(result_reverse, reverse.c_str(), reverse.length(), control.c_str(), control.length(), matrix);
    char* cigar_decoded = parasail_cigar_decode(cigar);
    
    if(result->score > result_reverse->score) {
	
        complete_result.score = result->score;
        complete_result.ref = traceback ->ref;
        complete_result.comp = traceback->comp;
        complete_result.query = traceback->query;
        complete_result.orientation = '+';
        complete_result.cigar = cigar_decoded;
    }
    else {
        complete_result.score = result_reverse->score;
        complete_result.ref = traceback_reverse ->ref;
        complete_result.comp = traceback_reverse->comp;
        complete_result.query = traceback_reverse->query;
        complete_result.orientation = '-';
        complete_result.cigar = cigar_decoded;
    }
    parasail_traceback_free(traceback);
    parasail_result_free(result);
    parasail_traceback_free(traceback_reverse);
    parasail_result_free(result_reverse);

    return complete_result;
}

// Get a map of all positions for reads in bam, for each window get the read IDs and store id in the sample dict
std::unordered_map<std::string, std::map<int, std::vector<std::pair<std::string,std::string>>>> get_reads_in_window(std::unordered_map<std::string, std::string>& samples_to_bam, std::unordered_map<std::string, std::map<int,int>>& positions, int window){
	using namespace std;
	unordered_map<string, map<int, vector<pair<string,string>>>> positions_to_reads;
	for(unordered_map<string, string>::iterator it = samples_to_bam.begin(); it != samples_to_bam.end(); ++it){
		string sample = it.first;
		string bam = it.second;
		// Open up the bam
		auto bam_file = sam_open(bam.c_str(), file_type.c_str());
		auto bam_header = sam_hdr_read(bam_file);
		auto bam_record = bam_init1();
		while (sam_read1(bam_file, bam_header, bam_record) >= 0){
			// Compute the reference positions
			int start = (int) bam_record->core.pos - window;
			string chrom = string(bam_header->target_name[aln->core.tid]);
			auto cigar = bam_get_cigar(bam_record);
			string q_name = string(bam_get_qname(bam_record));
			int end = start;
			for (int k = 0; k < bam_record->core.n_cigar; k++){
				int op = bam_cigar_op(cigar[k]);
				int ol = bam_cigar_oplen(cigar[k]);
				if (op == BAM_CMATCH || op == BAM_CDEL || op == BAM_CREF_SKIP || op == BAM_CEQUAL || op == BAM_CDIFF){
					end += ol;
				}
			}
			end += window;
			// Need to now check if there are any entries within positions between start and end
			auto lower = positions[chrom].lower_bound(start);
    		auto upper = positions[chrom].upper_bound(end);
    		if(lower != upper){
    			// have a hit
    			for(auto it2 = lower; it2 != upper; ++it2){
    				int window = it2.first;
    				positions_to_reads[chrom][window].push_back(make_pair(q_name, sample));
    			}
    		}
		}
	}
	return positions_to_reads;
}


// For now void, will return some number of maps later
std::unordered_map<std::string, std::string> get_hap_set(std::string& chrom, int start, int window, std::vector<std::pair<std::string, std::string>>& reads_in_window, std::unordered_map<std::string, std::string>& samples_to_fastq, std::string& ref_sequence, std::unordered_map<std::string, std::vector<TsvRecord>>& inserts_per_sample, int ref_window){
	// For each insert in window, splice insert into ref sequence to create haplotype
	using namespace std;
	unordered_map<string, string> hap_set; 
	for(unordered_map<string, vector<TsvRecord>>::iterator it = inserts_per_sample.begin(); it != inserts_per_sample.end(); ++it){
		string sample = it.first;
		for(size_t i = 0; i < it.second.size(); i++){
			// Compute insertion position on ref seq and splice in insertion
			int ref_start = start - ref_window;
			int insert_start = it.second[i].ref_start - ref_start;
			int insert_end = it.second[i].ref_end - ref_start;
			// Insert the sequence at insert_start
			string hap = ref_sequence.substr(0, insert_start) + it.second[i].seq + ref_sequence.substr(insert_end);
			hap_set[it.second[i].read_id] = hap;
		}
	}
	return hap_set;
}

std::string get_best_hap(std::unordered_map<std::string, std::vector<TsvRecord>>& inserts_per_sample, std::unordered_map<std::string, std::string>& hap_set){
	// For each insert in inserts_per_sample, align it to each hap
	using namespace std;
	unordered_map<string, int> scores;
	for(unordered_map<string, vector<TsvRecord>>::iterator it = inserts_per_sample.begin(); it != inserts_per_sample.end(); ++it){
		string sample = it.first;
		auto read_index = fai_load(samples_to_fastq[sample].c_str());
		for(size_t i = 0; i < it.second.size(); i++){
			// Compute insertion position on ref seq and splice in insertion
			char* tmp_read = fai_fetch(read_index,it.second[i].read_id.c_str());
			string read_sequence = string(tmp_read);
			for(unordered_map<string,string>::iterator it2 = hap_set.begin(); it2 != hap_set.end(); it2++){
				string hap_sequence = it2.second;
				AlignmentResult result = align_parasail(read_sequence, hap_sequence, matrix);
				scores[it2.first] += result.score;
			}
			free(tmp_read);
		}
	}
	// Compute and return the highest scoring hap
	int best_score = 0;
	string best_hap = "";
	for(unordered_map<string, int>::iterator it = scores.begin(); it != scores.end(); it++){
		if(it.second > best_score){
			best_score = it.second;
			best_hap = it.first;
		}
	}
	return best_hap;
}

void ReAlign::realign_all(std::string& bam_list, std::string& tsv_list, std::string& fastq_list, std::string& sample_list, std::string& output_tsv, std::string& output_bam, int window, std::string& ref, int ref_window){
	using namespace std;
	// Read in the lists
	vector<string> samples;
	vector<string> bams;
	vector<string> tsvs;
	vector<string> fastqs;
	stringstream sss(sample_list);
	for (int i; sss >> i;) {
		samples.push_back(i);
		if (sss.peek() == ',')
			sss.ignore();
	}
	unordered_map<string, string> samples_to_bam;
	stringstream ssb(bam_list);
	int count = 0;
	for (int i; ssb >> i;) {
		bams.push_back(i);
		samples_to_bam[samples[count]] = i;
		count++;
		if (ssb.peek() == ',')
			ssb.ignore();
	}
	unordered_map<string, string> samples_to_fastq;
	stringstream ssf(fastq_list);
	int count = 0;
	for (int i; ssf >> i;) {
		fastqs.push_back(i);
		samples_to_fastq[samples[count]] = i;
		count++;
		if (ssf.peek() == ',')
			ssf.ignore();
	}
	unordered_map<string, string> samples_to_tsv;
	stringstream sst(tsv_list);
	int count = 0;
	for (int i; sst >> i;) {
		tsvs.push_back(i);
		samples_to_tsv[samples[count]] = i;
		count++;
		if (sst.peek() == ',')
			sst.ignore();
	}
	// Open each tsv and read in data
	all_tsv_records = unordered_map<string, map<int, vector<TsvRecord>>>;
	positions_per_chrom = unordered_map<string, pair<int,int>>;
	for(unordered_map<string, string>::iterator it = samples_to_tsv.begin(); it != samples_to_tsv.end(); ++it){
		string sample = it.first;
		string tsv_file = it.second;
		// Open and read tsv file
		ifstream infile(it.second);
		string chrom, read, seq, annotation;
		int ref_start, ref_end, read_start, read_end, merged
		// First line is header
		getline(infile, chrom);
		while(infile >> chrom >> ref_start >> ref_end >> read >> read_start >> read_end >> merged >> seq >> annotation){
			// For each line create a TSV record class that has everything set correctly if it is passing
			if(annotation != "PASS"){
				continue;
			}
			// Setup tsv record
			r = TsvRecord(read, read_start, read_end, chrom, ref_start, ref_end, merged, seq, sample);
			all_tsv_records[chrom][ref_start].push_back(r);
			if(positions_per_chrom.count(chrom) == 0){
				positions_per_chrom[chrom] = make_pair(numeric_limits<int>::max(), 0);
			}
			if(ref_start < positions_per_chrom[chrom].first){
				positions_per_chrom[chrom].first = ref_start;
			}
			if(ref_end > positions_per_chrom[chrom].second){
				positions_per_chrom[chrom].second = ref_end;
			}

		}
	}
	// Iterate through each window and get a vector of all tsv_records in that window
	unordered_map<string, map<int,int>> positions;
	for(unordered_map<string, map<int, vector<TsvRecord>>>::iterator it = all_tsv_records.begin(); it != all_tsv_records.end(); ++it){
		string chrom = it.first;
		map<int,int> tmp;
		positions[chrom] = tmp;
		// TODO : Get Reference sequence from the ref
		ref_sequence = "";
		int count = positions_per_chrom[chrom].first;
		while(count < positions_per_chrom[chrom].second){
			auto lower = it.second.lower_bound(count);
    		auto upper = it.second.upper_bound(count+window);
    		if(lower != upper){
    			positions[chrom][count] = 1;
    		}
    		count += window;
    	}
    }

    // Setup faidx for reference
    auto ref_index = fai_load(ref.c_str());

    // Pass these positions to get a list of reads/samples per window
    unordered_map<string, map<int, vector<pair<string,string>>>> reads_in_window = get_reads_in_window(samples_to_bam, positions, window);
    // For each window now that we have all reads, get the haplotypes, then align haplotypes to insert supporting reads, and then 
    for(unordered_map<string, map<int, vector<TsvRecord>>>::iterator it = all_tsv_records.begin(); it != all_tsv_records.end(); ++it){
		string chrom = it.first;
		map<int,int> tmp;
		positions[chrom] = tmp;
		// TODO : Get Reference sequence from the ref
		char* tmp_ref = fai_fetch(ref_index,chrom.c_str());
		string ref_sequence = string(tmp_ref);
		int count = positions_per_chrom[chrom].first;
		while(count < positions_per_chrom[chrom].second){
			auto lower = it.second.lower_bound(count);
    		auto upper = it.second.upper_bound(count+window);
    		if(lower != upper){
    			// Have at least one insert in this range
    			// Setup a list per sample
    			inserts_per_sample = unordered_map<string, vector<TsvRecord>>;
    			for(auto it2 = lower; it2 != upper; ++it2){
    				for(int j = 0; j < it2.second.size(); j++){
    					inserts_per_sample[it2.second[j].sample].push_back(it2.second[j]);
    				}
    			}
    			// Run re-align on the window
    			unordered_map<string, string> hap_set = get_hap_set(chrom, count, window, reads_in_window[chrom][count], samples_to_fastq, ref_sequence, inserts_per_sample, ref_window);
    			// Take hap sets and align it to each read with an insertion, Select best haplotype
    			string best_hap= get_best_haps(reads_in_window[chrom][count], hap_set);
    			// Align reads to best haplotype and ref. If read aligns better to haplotype than ref, update record
    			unorderd_map<string, vector<string>> reads_per_sample;
    			for(size_t i = 0; i < reads_in_window[chrom][count].size(); i++){
    				reads_per_sample[reads_in_window[chrom][count][i].first].push_back(reads_in_window[chrom][count][i].second);
    			}
    			AlignmentResult hap_to_ref = align_parasail(best_hap, ref_sequence, matrix);
    			for(unordered_map<string, vector<string>>::iterator it2 = reads_per_sample.begin(); it2 != reads_per_sample.end(); ++it2){
    				auto read_index = fai_load(samples_to_fastq[it2.first].c_str());
    				for(size_t i = 0; i < it2.second.size(); i++){
    					char* tmp_read = fai_fetch(read_index, it2.second[i].c_str());
						string read_sequence = string(tmp_read);
						AlignmentResult ref_result = align_parasail(read_sequence, ref_sequence, matrix);
						AlignmentResult alt_result = align_parasail(read_sequence, best_hap, matrix);
						free(tmp_read);
						if(alt_result.score > ref_result.score){
							// Need to output a tsv record here
						}
    				}
    			}
    		}
    		count += window;
    	}
    	free(tmp_ref);
    }
}



















